### PR TITLE
fix RPIT ICE for implicit HRTB when missing dyn

### LINKED ISSUE #101297: ICE: generic RPIT with implicit HRTB when missing dyn
### Code
Found originally via https://discord.com/channels/273534239310479360/273541522815713281/1014981030022750328

I managed to reduce it to
```rust
fn ice() -> impl AsRef<Fn(&())> {
    todo!()
}
```

[playground link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=66591852861d8334fe46a9f3cb9417df)

This seems to be related to a combination of RPIT, missing `dyn` keyword, implicit HRTB (`&()`) and parenthesis syntax for the inner trait

### Meta

ICEs on both current beta (1.64) and latest nightly (1.65), but **not** on stable (1.63)

`rustc --version --verbose`:
```
rustc 1.65.0-nightly (bc4b39c27 2022-08-29)
binary: rustc
commit-hash: bc4b39c271bbd36736cbf1c0a1ac23d5df38d365
commit-date: 2022-08-29
host: x86_64-pc-windows-msvc
release: 1.65.0-nightly
LLVM version: 15.0.0
```

### Error output

```
thread 'rustc' panicked at 'no entry for node id: `NodeId(49)`', compiler\rustc_ast_lowering\src\lib.rs:526:55

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.65.0-nightly (bc4b39c27 2022-08-29) running on x86_64-pc-windows-msvc

note: compiler flags: --crate-type bin -C embed-bitcode=no -C debuginfo=2 -C incremental     

note: some of the compiler flags provided by cargo are hidden
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
stack backtrace:
   0:     0x7ff83e138c76 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h6bde120db1d6d2fe
   1:     0x7ff83e17414b - core::fmt::write::h77aade6c39e8b116
   2:     0x7ff83e12b7ea - <std::io::IoSlice as core::fmt::Debug>::fmt::h101d3289d91f656e    
   3:     0x7ff83e13c234 - std::panicking::default_hook::h580f6f477c6eceb4
   4:     0x7ff83e13be6a - std::panicking::default_hook::h580f6f477c6eceb4
   5:     0x7ff82a7261ee - rustc_driver[fd6bcdea86b534bf]::describe_lints
   6:     0x7ff83e13cc32 - std::panicking::rust_panic_with_hook::hc6a9e3dedab4cd8f
   7:     0x7ff83e13c9bd - <std::panicking::begin_panic_handler::StrPanicPayload as core::panic::BoxMeUp>::get::h38f790290b2cbd88
   8:     0x7ff83e13997f - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h6bde120db1d6d2fe
   9:     0x7ff83e13c680 - rust_begin_unwind
  10:     0x7ff83e1a94e5 - core::panicking::panic_fmt::hdcb6d460fcce0538
  11:     0x7ff8287611d5 - <rustc_ast_lowering[36e57923481bc2ad]::index_crate::Indexer as rustc_ast[74432c11d4c10689]::visit::Visitor>::visit_foreign_item
  12:     0x7ff8287666ed - <rustc_ast_lowering[36e57923481bc2ad]::index_crate::Indexer as rustc_ast[74432c11d4c10689]::visit::Visitor>::visit_foreign_item
  13:     0x7ff829a709ce - rustc_ast_lowering[36e57923481bc2ad]::lower_to_hir
  14:     0x7ff829a87387 - rustc_ast_lowering[36e57923481bc2ad]::lower_to_hir
  15:     0x7ff829a5ac2e - <rustc_mir_transform[fc712f60699b1a62]::generator::StateTransform 
as rustc_middle[4358913ddde1411f]::mir::MirPass>::run_pass
  16:     0x7ff829a7f0ff - rustc_ast_lowering[36e57923481bc2ad]::lower_to_hir
  17:     0x7ff829adfae4 - <rustc_ast_lowering[36e57923481bc2ad]::index::NodeCollector as rustc_hir[18bcb375ee80570a]::intravisit::Visitor>::visit_item
  18:     0x7ff829a686e1 - rustc_ast_lowering[36e57923481bc2ad]::lower_to_hir
  19:     0x7ff828a687e5 - <&rustc_index[3755210453515e92]::vec::IndexVec<rustc_middle[4358913ddde1411f]::mir::Promoted, rustc_middle[4358913ddde1411f]::mir::Body> as rustc_serialize[718946734402e014]::serialize::Decodable<rustc_query_impl[772085091de65120]::on_disk_cache::CacheDecoder>>::decode
  20:     0x7ff828bd061f - <&[(rustc_middle[4358913ddde1411f]::middle::exported_symbols::ExportedSymbol, rustc_middle[4358913ddde1411f]::middle::exported_symbols::SymbolExportInfo)] as rustc_serialize[718946734402e014]::serialize::Decodable<rustc_query_impl[772085091de65120]::on_disk_cache::CacheDecoder>>::decode
  21:     0x7ff828acb878 - <&[(rustc_middle[4358913ddde1411f]::middle::exported_symbols::ExportedSymbol, rustc_middle[4358913ddde1411f]::middle::exported_symbols::SymbolExportInfo)] as rustc_serialize[718946734402e014]::serialize::Decodable<rustc_query_impl[772085091de65120]::on_disk_cache::CacheDecoder>>::decode
  22:     0x7ff828b9f30f - <&[(rustc_middle[4358913ddde1411f]::middle::exported_symbols::ExportedSymbol, rustc_middle[4358913ddde1411f]::middle::exported_symbols::SymbolExportInfo)] as rustc_serialize[718946734402e014]::serialize::Decodable<rustc_query_impl[772085091de65120]::on_disk_cache::CacheDecoder>>::decode
  23:     0x7ff82a5ee45d - <rustc_middle[4358913ddde1411f]::ty::rvalue_scopes::RvalueScopes>::temporary_scope
  24:     0x7ff82a1921f3 - <rustc_span[494f2df5f00b7bc9]::symbol::Symbol as rustc_serialize[718946734402e014]::serialize::Encodable<rustc_query_impl[772085091de65120]::on_disk_cache::CacheEncoder>>::encode
  25:     0x7ff82a175d67 - <rustc_span[494f2df5f00b7bc9]::symbol::Symbol as rustc_serialize[718946734402e014]::serialize::Encodable<rustc_query_impl[772085091de65120]::on_disk_cache::CacheEncoder>>::encode
  26:     0x7ff82a61ba5c - <rustc_middle[4358913ddde1411f]::hir::map::Map>::attrs
  27:     0x7ff8289006c1 - <rustc_passes[1bfdf1db058ebff1]::dead::MarkSymbolVisitor as rustc_hir[18bcb375ee80570a]::intravisit::Visitor>::visit_anon_const
  28:     0x7ff82c2520e0 - <rustc_query_impl[772085091de65120]::queries::permits_zero_init as rustc_query_system[52e386d7ed70a530]::query::config::QueryDescription<rustc_query_impl[772085091de65120]::plumbing::QueryCtxt>>::describe
  29:     0x7ff82c394526 - <&[rustc_middle[4358913ddde1411f]::ty::abstract_const::Node] as rustc_serialize[718946734402e014]::serialize::Decodable<rustc_query_impl[772085091de65120]::on_disk_cache::CacheDecoder>>::decode
  30:     0x7ff828b15a75 - <&[(rustc_middle[4358913ddde1411f]::middle::exported_symbols::ExportedSymbol, rustc_middle[4358913ddde1411f]::middle::exported_symbols::SymbolExportInfo)] as rustc_serialize[718946734402e014]::serialize::Decodable<rustc_query_impl[772085091de65120]::on_disk_cache::CacheDecoder>>::decode
  31:     0x7ff828b9ef17 - <&[(rustc_middle[4358913ddde1411f]::middle::exported_symbols::ExportedSymbol, rustc_middle[4358913ddde1411f]::middle::exported_symbols::SymbolExportInfo)] as rustc_serialize[718946734402e014]::serialize::Decodable<rustc_query_impl[772085091de65120]::on_disk_cache::CacheDecoder>>::decode
  32:     0x7ff828a22d81 - <rustc_query_impl[772085091de65120]::Queries as rustc_middle[4358913ddde1411f]::ty::query::QueryEngine>::try_mark_green
  33:     0x7ff82802eb71 - rustc_interface[54e1975349ddcc95]::util::commit_date_str
  34:     0x7ff828034def - rustc_interface[54e1975349ddcc95]::util::commit_date_str
  35:     0x7ff8280410cd - rustc_interface[54e1975349ddcc95]::passes::analysis
  36:     0x7ff828a6869e - <&rustc_index[3755210453515e92]::vec::IndexVec<rustc_middle[4358913ddde1411f]::mir::Promoted, rustc_middle[4358913ddde1411f]::mir::Body> as rustc_serialize[718946734402e014]::serialize::Decodable<rustc_query_impl[772085091de65120]::on_disk_cache::CacheDecoder>>::decode
  37:     0x7ff828bce163 - <&[(rustc_middle[4358913ddde1411f]::middle::exported_symbols::ExportedSymbol, rustc_middle[4358913ddde1411f]::middle::exported_symbols::SymbolExportInfo)] as rustc_serialize[718946734402e014]::serialize::Decodable<rustc_query_impl[772085091de65120]::on_disk_cache::CacheDecoder>>::decode
  38:     0x7ff828b16cf2 - <&[(rustc_middle[4358913ddde1411f]::middle::exported_symbols::ExportedSymbol, rustc_middle[4358913ddde1411f]::middle::exported_symbols::SymbolExportInfo)] as rustc_serialize[718946734402e014]::serialize::Decodable<rustc_query_impl[772085091de65120]::on_disk_cache::CacheDecoder>>::decode
  39:     0x7ff828b9ec3f - <&[(rustc_middle[4358913ddde1411f]::middle::exported_symbols::ExportedSymbol, rustc_middle[4358913ddde1411f]::middle::exported_symbols::SymbolExportInfo)] as rustc_serialize[718946734402e014]::serialize::Decodable<rustc_query_impl[772085091de65120]::on_disk_cache::CacheDecoder>>::decode
  40:     0x7ff828007036 - <unknown>
  41:     0x7ff827fe6428 - <unknown>
  42:     0x7ff827fe2c1a - <unknown>
  43:     0x7ff827fe75d3 - <unknown>
  44:     0x7ff828017907 - <rustc_middle[4358913ddde1411f]::ty::SymbolName as core[dd21b54d167c7b0b]::fmt::Display>::fmt
  45:     0x7ff828007809 - <unknown>
  46:     0x7ff828007bbd - <unknown>
  47:     0x7ff83e14d63c - std::sys::windows::thread::Thread::new::h0a36a13d1ba46f75
  48:     0x7ff8d1a57034 - BaseThreadInitThunk
  49:     0x7ff8d1e42651 - RtlUserThreadStart
  
query stack during panic:
thread 'rustc' panicked at 'called `Option::unwrap()` on a `None` value', /rustc/bc4b39c271bbd36736cbf1c0a1ac23d5df38d365\compiler\rustc_query_system\src\query\plumbing.rs:341:11        
stack backtrace:
   0:     0x7ff83e138c76 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h6bde120db1d6d2fe
   1:     0x7ff83e17414b - core::fmt::write::h77aade6c39e8b116
   2:     0x7ff83e12b7ea - <std::io::IoSlice as core::fmt::Debug>::fmt::h101d3289d91f656e    
   3:     0x7ff83e13c234 - std::panicking::default_hook::h580f6f477c6eceb4
   4:     0x7ff83e13be6a - std::panicking::default_hook::h580f6f477c6eceb4
   5:     0x7ff82a7261ee - rustc_driver[fd6bcdea86b534bf]::describe_lints
   6:     0x7ff83e13cc32 - std::panicking::rust_panic_with_hook::hc6a9e3dedab4cd8f
   7:     0x7ff83e13c97a - <std::panicking::begin_panic_handler::StrPanicPayload as core::panic::BoxMeUp>::get::h38f790290b2cbd88
   8:     0x7ff83e13997f - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h6bde120db1d6d2fe
   9:     0x7ff83e13c680 - rust_begin_unwind
  10:     0x7ff83e1a94e5 - core::panicking::panic_fmt::hdcb6d460fcce0538
  11:     0x7ff83e1a938c - core::panicking::panic::h461763e65d8957b4
  12:     0x7ff82a17472b - <rustc_span[494f2df5f00b7bc9]::symbol::Symbol as rustc_serialize[718946734402e014]::serialize::Encodable<rustc_query_impl[772085091de65120]::on_disk_cache::CacheEncoder>>::encode
  13:     0x7ff82a057507 - <rustc_query_impl[772085091de65120]::Queries as rustc_middle[4358913ddde1411f]::ty::query::QueryEngine>::as_any
  14:     0x7ff82c21cf77 - <rustc_span[494f2df5f00b7bc9]::def_id::DefId as rustc_query_impl[772085091de65120]::keys::Key>::default_span
  15:     0x7ff82c200a94 - <rustc_query_impl[772085091de65120]::plumbing::QueryCtxt>::try_print_query_stack
  16:     0x7ff82c2bbc65 - <&[rustc_middle[4358913ddde1411f]::ty::abstract_const::Node] as rustc_serialize[718946734402e014]::serialize::Decodable<rustc_query_impl[772085091de65120]::on_disk_cache::CacheDecoder>>::decode
  17:     0x7ff82c2f2ad6 - <&[rustc_middle[4358913ddde1411f]::ty::abstract_const::Node] as rustc_serialize[718946734402e014]::serialize::Decodable<rustc_query_impl[772085091de65120]::on_disk_cache::CacheDecoder>>::decode
  18:     0x7ff82c1e6fec - <alloc[e6e4c83829d8bf3c]::vec::Vec<u8> as rustc_serialize[718946734402e014]::serialize::Decodable<rustc_query_impl[772085091de65120]::on_disk_cache::CacheDecoder>>::decode
  19:     0x7ff828acba57 - <&[(rustc_middle[4358913ddde1411f]::middle::exported_symbols::ExportedSymbol, rustc_middle[4358913ddde1411f]::middle::exported_symbols::SymbolExportInfo)] as rustc_serialize[718946734402e014]::serialize::Decodable<rustc_query_impl[772085091de65120]::on_disk_cache::CacheDecoder>>::decode
  20:     0x7ff828b9f30f - <&[(rustc_middle[4358913ddde1411f]::middle::exported_symbols::ExportedSymbol, rustc_middle[4358913ddde1411f]::middle::exported_symbols::SymbolExportInfo)] as rustc_serialize[718946734402e014]::serialize::Decodable<rustc_query_impl[772085091de65120]::on_disk_cache::CacheDecoder>>::decode
  21:     0x7ff82a5ed677 - <rustc_middle[4358913ddde1411f]::ty::rvalue_scopes::RvalueScopes>::temporary_scope
  22:     0x7ff82a18f742 - <rustc_span[494f2df5f00b7bc9]::symbol::Symbol as rustc_serialize[718946734402e014]::serialize::Encodable<rustc_query_impl[772085091de65120]::on_disk_cache::CacheEncoder>>::encode
  23:     0x7ff82a157e2c - <rustc_span[494f2df5f00b7bc9]::symbol::Symbol as rustc_serialize[718946734402e014]::serialize::Encodable<rustc_query_impl[772085091de65120]::on_disk_cache::CacheEncoder>>::encode
  24:     0x7ff82a5ee700 - <rustc_middle[4358913ddde1411f]::ty::rvalue_scopes::RvalueScopes>::temporary_scope
  25:     0x7ff82a197c30 - <rustc_span[494f2df5f00b7bc9]::symbol::Symbol as rustc_serialize[718946734402e014]::serialize::Encodable<rustc_query_impl[772085091de65120]::on_disk_cache::CacheEncoder>>::encode
  26:     0x7ff82a1722f6 - <rustc_span[494f2df5f00b7bc9]::symbol::Symbol as rustc_serialize[718946734402e014]::serialize::Encodable<rustc_query_impl[772085091de65120]::on_disk_cache::CacheEncoder>>::encode
  27:     0x7ff82a057507 - <rustc_query_impl[772085091de65120]::Queries as rustc_middle[4358913ddde1411f]::ty::query::QueryEngine>::as_any
  28:     0x7ff82c21cf77 - <rustc_span[494f2df5f00b7bc9]::def_id::DefId as rustc_query_impl[772085091de65120]::keys::Key>::default_span
  29:     0x7ff82c200a94 - <rustc_query_impl[772085091de65120]::plumbing::QueryCtxt>::try_print_query_stack
  30:     0x7ff82c2bbd25 - <&[rustc_middle[4358913ddde1411f]::ty::abstract_const::Node] as rustc_serialize[718946734402e014]::serialize::Decodable<rustc_query_impl[772085091de65120]::on_disk_cache::CacheDecoder>>::decode
  31:     0x7ff82c2f2ad6 - <&[rustc_middle[4358913ddde1411f]::ty::abstract_const::Node] as rustc_serialize[718946734402e014]::serialize::Decodable<rustc_query_impl[772085091de65120]::on_disk_cache::CacheDecoder>>::decode
  32:     0x7ff82c1e705e - <alloc[e6e4c83829d8bf3c]::vec::Vec<u8> as rustc_serialize[718946734402e014]::serialize::Decodable<rustc_query_impl[772085091de65120]::on_disk_cache::CacheDecoder>>::decode
  33:     0x7ff82c27818c - <(rustc_middle[4358913ddde1411f]::ty::instance::Instance, rustc_span[494f2df5f00b7bc9]::def_id::LocalDefId) as rustc_query_impl[772085091de65120]::keys::Key>::default_span
  34:     0x7ff82a79c84d - rustc_interface[54e1975349ddcc95]::interface::try_print_query_stack
  35:     0x7ff82a726ed8 - rustc_driver[fd6bcdea86b534bf]::report_ice
  36:     0x7ff83e13cc32 - std::panicking::rust_panic_with_hook::hc6a9e3dedab4cd8f
  37:     0x7ff83e13c9bd - <std::panicking::begin_panic_handler::StrPanicPayload as core::panic::BoxMeUp>::get::h38f790290b2cbd88
  38:     0x7ff83e13997f - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h6bde120db1d6d2fe
  39:     0x7ff83e13c680 - rust_begin_unwind
  40:     0x7ff83e1a94e5 - core::panicking::panic_fmt::hdcb6d460fcce0538
  41:     0x7ff8287611d5 - <rustc_ast_lowering[36e57923481bc2ad]::index_crate::Indexer as rustc_ast[74432c11d4c10689]::visit::Visitor>::visit_foreign_item
  42:     0x7ff8287666ed - <rustc_ast_lowering[36e57923481bc2ad]::index_crate::Indexer as rustc_ast[74432c11d4c10689]::visit::Visitor>::visit_foreign_item
  43:     0x7ff829a709ce - rustc_ast_lowering[36e57923481bc2ad]::lower_to_hir
  44:     0x7ff829a87387 - rustc_ast_lowering[36e57923481bc2ad]::lower_to_hir
  45:     0x7ff829a5ac2e - <rustc_mir_transform[fc712f60699b1a62]::generator::StateTransform 
as rustc_middle[4358913ddde1411f]::mir::MirPass>::run_pass
  46:     0x7ff829a7f0ff - rustc_ast_lowering[36e57923481bc2ad]::lower_to_hir
  47:     0x7ff829adfae4 - <rustc_ast_lowering[36e57923481bc2ad]::index::NodeCollector as rustc_hir[18bcb375ee80570a]::intravisit::Visitor>::visit_item
  48:     0x7ff829a686e1 - rustc_ast_lowering[36e57923481bc2ad]::lower_to_hir
  49:     0x7ff828a687e5 - <&rustc_index[3755210453515e92]::vec::IndexVec<rustc_middle[4358913ddde1411f]::mir::Promoted, rustc_middle[4358913ddde1411f]::mir::Body> as rustc_serialize[718946734402e014]::serialize::Decodable<rustc_query_impl[772085091de65120]::on_disk_cache::CacheDecoder>>::decode
  50:     0x7ff828bd061f - <&[(rustc_middle[4358913ddde1411f]::middle::exported_symbols::ExportedSymbol, rustc_middle[4358913ddde1411f]::middle::exported_symbols::SymbolExportInfo)] as rustc_serialize[718946734402e014]::serialize::Decodable<rustc_query_impl[772085091de65120]::on_disk_cache::CacheDecoder>>::decode
  51:     0x7ff828acb878 - <&[(rustc_middle[4358913ddde1411f]::middle::exported_symbols::ExportedSymbol, rustc_middle[4358913ddde1411f]::middle::exported_symbols::SymbolExportInfo)] as rustc_serialize[718946734402e014]::serialize::Decodable<rustc_query_impl[772085091de65120]::on_disk_cache::CacheDecoder>>::decode
  52:     0x7ff828b9f30f - <&[(rustc_middle[4358913ddde1411f]::middle::exported_symbols::ExportedSymbol, rustc_middle[4358913ddde1411f]::middle::exported_symbols::SymbolExportInfo)] as rustc_serialize[718946734402e014]::serialize::Decodable<rustc_query_impl[772085091de65120]::on_disk_cache::CacheDecoder>>::decode
  53:     0x7ff82a5ee45d - <rustc_middle[4358913ddde1411f]::ty::rvalue_scopes::RvalueScopes>::temporary_scope
  54:     0x7ff82a1921f3 - <rustc_span[494f2df5f00b7bc9]::symbol::Symbol as rustc_serialize[718946734402e014]::serialize::Encodable<rustc_query_impl[772085091de65120]::on_disk_cache::CacheEncoder>>::encode
  55:     0x7ff82a175d67 - <rustc_span[494f2df5f00b7bc9]::symbol::Symbol as rustc_serialize[718946734402e014]::serialize::Encodable<rustc_query_impl[772085091de65120]::on_disk_cache::CacheEncoder>>::encode
  56:     0x7ff82a61ba5c - <rustc_middle[4358913ddde1411f]::hir::map::Map>::attrs
  57:     0x7ff8289006c1 - <rustc_passes[1bfdf1db058ebff1]::dead::MarkSymbolVisitor as rustc_hir[18bcb375ee80570a]::intravisit::Visitor>::visit_anon_const
  58:     0x7ff82c2520e0 - <rustc_query_impl[772085091de65120]::queries::permits_zero_init as rustc_query_system[52e386d7ed70a530]::query::config::QueryDescription<rustc_query_impl[772085091de65120]::plumbing::QueryCtxt>>::describe
  59:     0x7ff82c394526 - <&[rustc_middle[4358913ddde1411f]::ty::abstract_const::Node] as rustc_serialize[718946734402e014]::serialize::Decodable<rustc_query_impl[772085091de65120]::on_disk_cache::CacheDecoder>>::decode
  60:     0x7ff828b15a75 - <&[(rustc_middle[4358913ddde1411f]::middle::exported_symbols::ExportedSymbol, rustc_middle[4358913ddde1411f]::middle::exported_symbols::SymbolExportInfo)] as rustc_serialize[718946734402e014]::serialize::Decodable<rustc_query_impl[772085091de65120]::on_disk_cache::CacheDecoder>>::decode
  61:     0x7ff828b9ef17 - <&[(rustc_middle[4358913ddde1411f]::middle::exported_symbols::ExportedSymbol, rustc_middle[4358913ddde1411f]::middle::exported_symbols::SymbolExportInfo)] as rustc_serialize[718946734402e014]::serialize::Decodable<rustc_query_impl[772085091de65120]::on_disk_cache::CacheDecoder>>::decode
  62:     0x7ff828a22d81 - <rustc_query_impl[772085091de65120]::Queries as rustc_middle[4358913ddde1411f]::ty::query::QueryEngine>::try_mark_green
  63:     0x7ff82802eb71 - rustc_interface[54e1975349ddcc95]::util::commit_date_str
  64:     0x7ff828034def - rustc_interface[54e1975349ddcc95]::util::commit_date_str
  65:     0x7ff8280410cd - rustc_interface[54e1975349ddcc95]::passes::analysis
  66:     0x7ff828a6869e - <&rustc_index[3755210453515e92]::vec::IndexVec<rustc_middle[4358913ddde1411f]::mir::Promoted, rustc_middle[4358913ddde1411f]::mir::Body> as rustc_serialize[718946734402e014]::serialize::Decodable<rustc_query_impl[772085091de65120]::on_disk_cache::CacheDecoder>>::decode
  67:     0x7ff828bce163 - <&[(rustc_middle[4358913ddde1411f]::middle::exported_symbols::ExportedSymbol, rustc_middle[4358913ddde1411f]::middle::exported_symbols::SymbolExportInfo)] as rustc_serialize[718946734402e014]::serialize::Decodable<rustc_query_impl[772085091de65120]::on_disk_cache::CacheDecoder>>::decode
  68:     0x7ff828b16cf2 - <&[(rustc_middle[4358913ddde1411f]::middle::exported_symbols::ExportedSymbol, rustc_middle[4358913ddde1411f]::middle::exported_symbols::SymbolExportInfo)] as rustc_serialize[718946734402e014]::serialize::Decodable<rustc_query_impl[772085091de65120]::on_disk_cache::CacheDecoder>>::decode
  69:     0x7ff828b9ec3f - <&[(rustc_middle[4358913ddde1411f]::middle::exported_symbols::ExportedSymbol, rustc_middle[4358913ddde1411f]::middle::exported_symbols::SymbolExportInfo)] as rustc_serialize[718946734402e014]::serialize::Decodable<rustc_query_impl[772085091de65120]::on_disk_cache::CacheDecoder>>::decode
  70:     0x7ff828007036 - <unknown>
  71:     0x7ff827fe6428 - <unknown>
  72:     0x7ff827fe2c1a - <unknown>
  73:     0x7ff827fe75d3 - <unknown>
  74:     0x7ff828017907 - <rustc_middle[4358913ddde1411f]::ty::SymbolName as core[dd21b54d167c7b0b]::fmt::Display>::fmt
  75:     0x7ff828007809 - <unknown>
  76:     0x7ff828007bbd - <unknown>
  77:     0x7ff83e14d63c - std::sys::windows::thread::Thread::new::h0a36a13d1ba46f75
  78:     0x7ff8d1a57034 - BaseThreadInitThunk
  79:     0x7ff8d1e42651 - RtlUserThreadStart
```

</p>
</details>

### PR DESCRIPTION
**Pull request by spastorino**
Closes #101297

r? @cjgillot 

cc @nikomatsakis

### CODE REVIEW

**Comment by spastorino** [comment]
This fixes a beta regression. I'm beta nominating this PR, given how minimal and no risk it is.

@rustbot label +beta-nominated

File: src/test/ui/impl-trait/generic-with-implicit-hrtb-without-dyn.rs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+// edition:2021
+
+fn ice() -> impl AsRef<Fn(&())> {
```
Comment (cjgillot) [inline src/test/ui/impl-trait/generic-with-implicit-hrtb-without-dyn.rs:3]: It also ICEs on edition 2015.  Could you add a `check-pass` test on edition 2015?

**Comment by apiraino** [comment]
Beta backport accepted as per compiler team [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202022-09-08/near/297790534)

@rustbot label +beta-accepted

**Pull request merged**